### ISSUE TITLE
Cannot change date range in overview

### ISSUE THREAD

**Issue by taniwallach** [projecthamster/hamster#639]
I installed on Debian buster using the current Git repository.

When hamster is open, I cannot "pull down" the date from the top menu to set a range of dates.

Navigation to different single days using the "<" and ">" icons does work.

I tried the snap also, and the had the same problem with it.

**Comment by GeraldJansen** [comment]
Hmm, strange. What do you mean by "pull down"? Just clicking on the date should open the popup. (I just tried current git on my Debian testing and it seems to work fine.)

**Comment by taniwallach** [comment]
There is a slight flicker, but nothing opens:
[Hamster-Debian10-Click-Date-no-reaction.zip](https://github.com/projecthamster/hamster/files/5544829/Hamster-Debian10-Click-Date-no-reaction.zip)

Maybe a missing depencency?

Maybe the "window service" is not starttng, as there was some mention of a need to sometimes kill 2 services, and I see only one:
```
tani@lxtani:/var/log$ ps -aux | grep hamster
tani     14975  0.0  0.2 301036 39012 ?        Sl   Nov15   0:08 python3 /usr/libexec/hamster/hamster-service
tani     15017  0.0  0.3 502800 60864 ?        Sl   Nov15   0:12 python3 /usr/bin/hamster --gapplication-service
tani     19992  0.0  0.0   7140   892 pts/1    S+   09:55   0:00 grep hamster
```

**Comment by GeraldJansen** [comment]
I also no longer see hamster-window-service running with recent versions. However, I do not get the --gapplication-service command line argument. You could try killing hamster and hamster-service, then from the command line try 
`/usr/bin/hamster version`, which should give v3.0.2-8-gceedcd87, and then try `/usr/bin/hamster &`.  

Sorry for the noise. Hopefully someone more knowledgeable about recent developments (eg. snap install) can help.

**Comment by taniwallach** [comment]
I think I have a fix in https://github.com/projecthamster/hamster/pull/640. It works for me. The "popup" opens/closes when the button is clicked. Click on the Today/Week/Month line, or on Apply makes the setting and closes the "popup" window.

The changes can be applied for testing on a Debian system by making the changes to `/usr/lib/python3/dist-packages/hamster/widgets/dates.py` and the running
```
pkill -f hamster-service ; /usr/bin/hamster
```

**Comment by taniwallach** [comment]
Note: I only briefly tried the snap to see if the issue did not occur there. All the main testing and attempts to find a patch were done on a version installed from Git. The patch was made to the installed code, and once it worked copies into the Git controlled directory to commit/push and make the PR.

```
tani@lxtani:~$ /usr/bin/hamster version
v3.0.2-8-gceedcd87
```

**Comment by matthijskooijman** [comment]
Interesting problem. I'm running Ubuntu Focal 20.04 and am not seeing this problem with the 3.0.2 Debian package, nor with git master.

Nice that the PR you submitted works, but it seems a bit like a shotgun approach, adding quite a bit of code (some of which seems pointless) which effectively seems to just disable the handling of `_hiding` (at least when `get_active()` is `True`).

Looking at the code, it seems likely that the `_hiding` attribute is somehow (incorrectly) preventing the popup from being shown. That attribute is intended to prevent a situation where the popup is shown, you click the button, then the popup hides because of the focus change, and then reshows again due to the button click immediately after.

I wonder if the problem is maybe that, on your system, the `toggle` event on the button triggers before the `on_focus_out` event or something, or there is some other reason that `_hiding` ends up as `True` when the popup is closed.

Does the popup *never* open? Or does it open once and then fail to open a second time?

Can you maybe try with this patch (without your PR applied), to see what events happen in what order?

```
--- a/src/hamster/widgets/dates.py
+++ b/src/hamster/widgets/dates.py
@@ -119,8 +119,10 @@ class RangePick(gtk.ToggleButton):
         self._ui = None
 
     def on_toggle(self, button):
+        print("on_toggle, active={}, _hiding={}".format(self.get_active(), self._hiding))
         if self.get_active():
             if self._hiding:
+                print("on_toggle, clearing _hiding")
                 self._hiding = False
                 self.set_active(False)
                 return
@@ -193,18 +195,22 @@ class RangePick(gtk.ToggleButton):
     def on_focus_out(self, popup, event):
         x, y = self.get_pointer()
         button_w, button_h = self.get_allocation().width, self.get_allocation().height
+        print("on_focus_out, active={}, _hiding={}, x={}, y={}, w={}, h={}".format(self.get_active(), self._hiding, x, y, button_w, button_h))
         # avoid double-toggling when focus goes from window to the toggle button
         if 0 <= x <= button_w and 0 <= y <= button_h:
+            print("on_focus_out, setting _hiding")
             self._hiding = True
 
         self.set_active(False)
 
 
     def hide(self):
+        print("hide, active={}, _hiding={}".format(self.get_active(), self._hiding))
         self.set_active(False)
         self.popup.hide()
 
     def show(self):
+        print("show, active={}, _hiding={}".format(self.get_active(), self._hiding))
         dummy, x, y = self.get_window().get_origin()
 
         alloc = self.get_allocation()
```

**Comment by taniwallach** [comment]
@matthijskooijman - Thanks for looking into this.

> Nice that the PR you submitted works, but it seems a bit like a shotgun approach, adding quite a bit of code (some of which seems pointless) which effectively seems to just disable the handling of `_hiding` (at least when `get_active()` is `True`).

Certainly. I'm not all that familiar with the hamser code, and am new to Python. I decided to see if I could determine where the issue/bug is and make an initial fix, so that eventually it could be done properly by someone who understands the code better.

> Does the popup _never_ open? Or does it open once and then fail to open a second time?

Before the patch - the popup never opened (or disappeared faster than it could be seen). No amount of clicking helped. My patch was fiddled with until it could open/close and would not get stuck open.

> Can you maybe try with this patch (without your PR applied), to see what events happen in what order?

Certainly. Here are the results.

After first click on the button (popup does not appear):
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
on_focus_out, active=True, _hiding=False, x=116, y=13, w=167, h=46
on_focus_out, setting _hiding
on_toggle, active=False, _hiding=True
hide, active=False, _hiding=True
```

After second click on the button (popup does not appear):
```
on_toggle, active=True, _hiding=True
on_toggle, clearing _hiding
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```

After third click on the button (popup does not appear) - seems the same as after the first click:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
on_focus_out, active=True, _hiding=False, x=79, y=15, w=167, h=46
on_focus_out, setting _hiding
on_toggle, active=False, _hiding=True
hide, active=False, _hiding=True
```

After forth click on the button (popup does not appear) - seems the same as after the first click:
```
on_toggle, active=True, _hiding=True
on_toggle, clearing _hiding
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```

Additional clicks seem to trigger the same behavior for odd numbered vs. even numbered clicks.

> Interesting problem. I'm running Ubuntu Focal 20.04 and am not seeing this problem with the 3.0.2 Debian package, nor with git master.

Then I'm guessing that some underlying package may be causing the issue on Debian 10, and whatever bug it was has already been fixed in the underlying package.

**Comment by taniwallach** [comment]
With this patch - things open and close fine.

Debugging output below.

---

Initial click:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=87, y=16, w=167, h=46
on_focus_out, skipping
```

after clicking on a different program window which also closed the popup:
```
on_focus_out, active=True, _hiding=False, x=-212, y=194, w=167, h=46
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```
---

Click to open again:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=111, y=21, w=167, h=46
on_focus_out, skipping
```
then clicked on "week":
```
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=196, y=116, w=196, h=46
```

---

Click to open again:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=113, y=17, w=201, h=46
on_focus_out, skipping
```
then clicked button again to close:
```
on_focus_out, active=True, _hiding=False, x=113, y=17, w=201, h=46
on_focus_out, skipping
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```

---

Several cycles of click to open + making a selection and it closing:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=65, y=18, w=199, h=46
on_focus_out, skipping
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=171, y=78, w=167, h=46


on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=93, y=17, w=167, h=46
on_focus_out, skipping
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=152, y=119, w=196, h=46


on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=98, y=30, w=196, h=46
on_focus_out, skipping
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=213, y=166, w=199, h=46


on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
on_focus_out, active=True, _hiding=False, x=131, y=26, w=199, h=46
on_focus_out, skipping
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=577, y=383, w=201, h=46
```

**Comment by matthijskooijman** [comment]
@taniwallach, I created a PR to fix this, could you doublecheck that it still works as expected for you and leave a review on the PR if it does?

**Comment by taniwallach** [comment]
I am working today also on a second machine, also with Debian 10. It does not have the bug about the popup not appearing fr the current master branch version.

However, on that machine - the popup always opens in the top left of the screen, and not placed over the hamster window (offset down and right a bit).

It is possible that the data below hints that these issues are related to one of:
  - `xorg` (was in use on the machine where there was an issue) vs `wayland` (in use on the machine where things worked) **seems more likely**
  - `ibus` (was in use on the machine where there was an issue)

---

The machine where I had the problem had the following shown by `printenv` (dropped the huge `LS_COLORS`):

```
tani@lxtani:~/Local/docker-hamster$ printenv | grep -Ev "LS_COLORS" | sort

CLUTTER_IM_MODULE=ibus
COLORTERM=truecolor
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
DESKTOP_SESSION=gnome-xorg
DISPLAY=:0
GDM_LANG=en_US.UTF-8
GDMSESSION=gnome-xorg
GJS_DEBUG_OUTPUT=stderr
GJS_DEBUG_TOPICS=JS ERROR;JS LOG
GNOME_DESKTOP_SESSION_ID=this-is-deprecated
GNOME_TERMINAL_SCREEN=/org/gnome/Terminal/screen/705cb0de_cd9f_4a58_9a73_38504368fce4
GNOME_TERMINAL_SERVICE=:1.483
GPG_AGENT_INFO=/run/user/1000/gnupg/S.gpg-agent:0:1
GSM_SKIP_SSH_AGENT_WORKAROUND=true
GTK_IM_MODULE=ibus
GTK_MODULES=gail:atk-bridge
HOME=/home/tani
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
LC_MEASUREMENT=en_IL
LC_MONETARY=en_IL
LC_NUMERIC=en_IL
LC_PAPER=en_IL
LC_TIME=en_IL
LOGNAME=tani
OLDPWD=/home/tani/Local
PATH=/home/tani/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/snap/bin
PWD=/home/tani/Local/docker-hamster
QT4_IM_MODULE=ibus
QT_ACCESSIBILITY=1
QT_IM_MODULE=ibus
SESSION_MANAGER=local/lxtani:@/tmp/.ICE-unix/2189,unix/lxtani:/tmp/.ICE-unix/2189
SHELL=/bin/bash
SHLVL=0
SSH_AUTH_SOCK=/run/user/1000/gnupg/S.gpg-agent.ssh
TERM=xterm-256color
USERNAME=tani
USER=tani
_=/usr/bin/printenv
VTE_VERSION=5402
WINDOWPATH=2
XAUTHORITY=/run/user/1000/gdm/Xauthority
XDG_CURRENT_DESKTOP=GNOME
XDG_DATA_DIRS=/usr/share/gnome:/home/tani/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop
XDG_MENU_PREFIX=gnome-
XDG_RUNTIME_DIR=/run/user/1000
XDG_SEAT=seat0
XDG_SESSION_CLASS=user
XDG_SESSION_DESKTOP=gnome-xorg
XDG_SESSION_ID=2
XDG_SESSION_TYPE=x11
XDG_VTNR=2
XMODIFIERS=@im=ibus

tani@lxtani:~/Local/docker-hamster$ 
```

The other machine where there is no problem has:
```
COLORTERM=truecolor
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
DESKTOP_SESSION=gnome
DISPLAY=:0
GDM_LANG=en_IL
GDMSESSION=gnome
GJS_DEBUG_OUTPUT=stderr
GJS_DEBUG_TOPICS=JS ERROR;JS LOG
GNOME_DESKTOP_SESSION_ID=this-is-deprecated
GNOME_TERMINAL_SCREEN=/org/gnome/Terminal/screen/93dcdfe2_5ace_418c_ae73_9516be452212
GNOME_TERMINAL_SERVICE=:1.61
GTK_MODULES=gail:atk-bridge
HOME=/home/tani
LANG=en_IL
LANGUAGE=en_IL:en
LOGNAME=tani
OLDPWD=/home/tani/Local/hamster-new
PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
PWD=/home/tani/Local/hamster-new/hamster
QT_ACCESSIBILITY=1
SESSION_MANAGER=local/lxtani:@/tmp/.ICE-unix/1565,unix/lxtani:/tmp/.ICE-unix/1565
SHELL=/bin/bash
SHLVL=0
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
TERM=xterm-256color
USERNAME=tani
USER=tani
_=/usr/bin/printenv
VTE_VERSION=5402
WAYLAND_DISPLAY=wayland-0
XDG_CURRENT_DESKTOP=GNOME
XDG_MENU_PREFIX=gnome-
XDG_RUNTIME_DIR=/run/user/1000
XDG_SEAT=seat0
XDG_SESSION_CLASS=user
XDG_SESSION_DESKTOP=gnome
XDG_SESSION_ID=2
XDG_SESSION_TYPE=wayland
XDG_VTNR=2
```

**Comment by matthijskooijman** [comment]
> xorg (was in use on the machine where there was an issue) vs wayland (in use on the machine where things worked) seems more likely
> ibus (was in use on the machine where there was an issue)

I'm using both xorg and ibus here (I think, based on `$QT_IM_MODULE`), without issues, so maybe it's something else.

But regardless, I think the PR I made just makes the code more robust regardless of what causes this particular event order, so IMHO it does not really matter what makes the difference.

**Comment by taniwallach** [comment]
Thanks. For the record - here is data I collected from the second machine:

---


Initial click:

```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
```

after clicking on a different program window which also closed the popup:
```
on_focus_out, active=True, _hiding=False, x=-92, y=13, w=167, h=46
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```

---

Click to open again:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
```

then clicked on "week":
```
hide, active=True, _hiding=False
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
on_focus_out, active=False, _hiding=False, x=67, y=42, w=196, h=46
```

---

Click to open again:
```
on_toggle, active=True, _hiding=False
show, active=True, _hiding=False
show done
```

then clicked button again to close:
```
on_focus_out, active=True, _hiding=False, x=144, y=17, w=199, h=46
on_focus_out, skipping
on_toggle, active=False, _hiding=False
hide, active=False, _hiding=False
```

**Comment by rhertzog** [comment]
Le mercredi 18 novembre 2020, Nathan Wallach a écrit :
> It is possible that the data below hints that these issues are related to one of:
>   - `xorg` (was in use on the machine where there was an issue) vs `wayland` (in use on the machine where things worked) **seems more likely**

I confirm that a combination of wayland + not very recent GTK version
creates this issue where some popups windows are misplaced. I got the same
problem in another application (tryton).

Cheers,
-- 
Raphaël Hertzog

**Comment by taniwallach** [comment]
> I confirm that a combination of wayland + not very recent GTK version creates this issue where some popups windows are misplaced. I got the same problem in another application (tryton). Cheers, -- Raphaël Hertzog

@rhertzog Thanks for the fast feedback. The placement issue (on machine running wayland)  is far less annoying than the missing popup. Debian is very conservative, so Debian 10 will probably remain on the older GTK until there is a Debian 11, but as it does not impair functionality and only convenience- I think user can live with the misplaced windows.

@matthijskooijman Thanks for all your effort in finding a good fix.

**Comment by GeraldJansen** [comment]
Some issues of popup positioning with wayland were fixed by #527. Perhaps the same fix is needed here.

**Comment by taniwallach** [comment]
> Some issues of popup positioning with wayland were fixed by #527. Perhaps the same fix is needed here.

Moved this separate issue and copied @GeraldJansen and @rhertzog posts to https://github.com/projecthamster/hamster/issues/645.

**Comment by matthijskooijman** [comment]
I'm just going to reopen this until a final fix is merged.

Thanks for opening up a new issue for the other problem, good to keep things separated.
